crates/hir/src/semantics.rs

```diff
@@ -225,7 +225,7 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         self.imp.type_of_pat(pat)
     }
 
-    pub fn type_of_pat_with_coercion(&self, expr: &ast::Pat) -> Option<Type> {
+    pub fn type_of_pat_with_coercion(&self, expr: &ast::Pat) -> Option<(Type, bool)> {
```

Comment:
The return type feels awkward to me -- complex structural types (nested tuples, enums) often are indicative that some concept is not named.  Could we would fold several functions here into just one

```rust
pub struct TypeInfo {
    ty: Type,
    coerced_from: Option<Type>
}

pub fn type_of() -> Option<TypeInfo>
```

if the answer is "yes", its better to land a refactor *after* PR.

Reply:
Yes I definitely agree, the return type isn't that descriptive. I'll explore that aferwards since we got a few more usages of this api now.